modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/AddCatalogWorkflowOperationHandler.java

```diff
@@ -0,0 +1,227 @@
+    CatalogTypeCollisionBehavior collBehavior  = parseCollisionBehavior(
+                           getCfgValueFromOp(wInst, CFG_KEY_CATALOG_TYPE_COLLISION_BEHAVIOR, false));
+
+    MediaPackage mp = wInst.getMediaPackage();
+
+    // if CatalogType is already part of the MediaPackage handle special cases
+    if (doesCatalogFlavorExist(catalogFlavor, mp.getCatalogs())) {
+      if (collBehavior == CatalogTypeCollisionBehavior.FAIL) {
+        throw new WorkflowOperationException("Catalog Type already exists and 'fail' was specified");
+      }
+      else if (collBehavior == CatalogTypeCollisionBehavior.SKIP) {
+        // don't add the Catalog
+        return createResult(mp, Action.CONTINUE);
+      }
+    }
+
+    // 'upload Catalog' to workspace
+    File   catalogFile = new File(catalogPath);
+    String catalogId   = UUID.randomUUID().toString();
+    URI    catalogURI  = null;
+    try (InputStream catalogInputStream = FileUtils.openInputStream(catalogFile)) {
+      catalogURI = workspace.put(mp.getIdentifier().toString(), catalogId,
+              catalogName, catalogInputStream);
+    }
+    catch (IOException e) {
+      throw new WorkflowOperationException(e);
+    }
+
+    // add Catalog to MediaPackage (and set Properties)
+    MediaPackageElement mpe = mp.add(catalogURI, MediaPackageElement.Type.Catalog, catalogFlavor);
+    mpe.setIdentifier(catalogId);
+    mpe.setMimeType(CATALOG_MIME_TYPE);
+    if (catalogTags != null) {
+      for (String tag : catalogTags.split(",")) {
+        mpe.addTag(tag);
+      }
+    }
+
+    return createResult(mp, Action.CONTINUE);
+  }
+
+  /**
+   * Checks whether the catalogFlavor exists in the array of catalogs
+   *
+   * @param catalogFlavor
+   * @param catalogs
+   * @return true, if the catalogFlavor exists in the array of catalogs, else false
+   */
+  private boolean doesCatalogFlavorExist(MediaPackageElementFlavor catalogFlavor, Catalog[] catalogs) {
+    List<Catalog> ctls = Arrays.asList(catalogs);
+
+    boolean doesFlavorExist = ctls.stream()
+      .anyMatch(cat -> catalogFlavor.matches(cat.getFlavor()));
+
+    return doesFlavorExist;
+  }
+
+  /**
+   * Gets a configuration value for a specific key from the current operation.
+   * isOptional can be used to specify whether null values are allowed to be returned
+   * or whether an WorkflowOperationException should be thrown.
+   *
+   * @param inst
+   *          used to get the value of the configuration
+   * @param cfgKey
+   *          specifies the value to get
+   * @param isOptional
+   *          is the configuration value allowed to be not set
+   * @return the value of the configuration key or null if isOptional is set and the configuration value is not set
+   * @throws WorkflowOperationException
+   *           if isOptional is false and the configuration value is not set
+   */
+  private String getCfgValueFromOp(WorkflowInstance inst, String cfgKey, boolean isOptional)
```

Comment:
Hmm.... this kind of code should go to WorkflowInstance or some Workflow utility class, shouldn't it? Also, have you checked that WorkflowInstance does not already offer similar methods?

Reply:
fixed it, the AbstractWorkflowOperationHandler provides similar methods